### PR TITLE
KOGITO-7826: Run Git and Storage services on a Web Worker

### PR DESCRIPTION
**Pull request by tiagobento**
This PR moves all Storage and Git services to a Web Worker (with the exception of DMN Runner Inputs, which run their own File System and is very lightweight).

This move was achieved by taking advantage of the fact that all these operations were already inside WorkspaceContextProvider, which was all Promise-based. More than that, by using a EnvelopeBusMessageManager both inside and outside the new Web Worker, all the communications between the main window and the workspaces worker is easily handled and type-checked. 10 points to ~Gryffindor~ Multiplying Architecture 🎉.

On small Workspaces (~100 files), the UX was already good, but now everything feels instantaneous.

On large Workspaces (~2.5k files, e.g. `kiegroup/kogito-examples`), the UX is now much better -- even good, I'd say -- with no more random reloads of the Editors during editing, and the Editors themselves running smoothly even when Storage and Git services are busy saving files or computing changes. No more choppy dragging when working with diagrams!

This also improves UX for slow machines, as the size of the Workspaces and slow CPUs both impact UX in the same way. The faster your CPU, the bigger the repository you can work with.

Both use-cases above were tested on a 2018 Mac Mini 32GB RAM and an M1 MacBook Air 8GB RAM, on Safari and Chrome.

---
### ⚠️ Keep in mind that working in such large Workspaces is NOT what KIE Sandbox was built for, so test it sparingly.

Throttling down the CPU in a small workspace is a better way of testing it.

A pre-built version of this branch is published here: https://tiagobento.github.io/kie-tools-staging/workspaces-on-web-worker/#/

---

Also on this PR:
- Removed URL-enabled JSON parser, as URL is not serializable, so can't be sent back and forth between main window and worker. Due to this, changed WorkspaceDescriptor origin URL to `string`.
- Created a new alert on the Editor page to indicate that a Commit is running.

### CODE REVIEW

File: packages/online-editor/src/workspace/worker/api/WorkspacesWorkerApi.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { WorkspacesWorkerGitApi } from "./WorkspacesWorkerGitApi";
+import { WorkspacesWorkerStorageApi } from "./WorkspacesWorkerStorageApi";
+
+export interface WorkspacesWorkerApi extends WorkspacesWorkerGitApi, WorkspacesWorkerStorageApi {
+  //
```
Comment (tomasdavidorg) [inline packages/online-editor/src/workspace/worker/api/WorkspacesWorkerApi.ts:21]: Is some comment missing here or this is just forgotten?

File: packages/online-editor/src/polyfill/BroadcastChannelSingleTabOnWorker.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      const subscription = subscriptions.get(name);
+      if (!subscription) {
+        subscriptions.set(name, new Set());
+      }
+
+      subscriptions.get(name)!.add(this);
+    }
+
+    public postMessage(message: any): void {
+      //FIXME: This could be one message per `setTimeout`, but it's fine for now.
```
Comment (tomasdavidorg) [inline packages/online-editor/src/polyfill/BroadcastChannelSingleTabOnWorker.ts:37]: Could we create a jira for this?

File: packages/online-editor/src/editor/FileSwitcher.tsx [hunk 3]
```diff
@@ -465,19 +463,20 @@ export function FileSvg(props: { workspaceFile: WorkspaceFile }) {
...
-              return;
-            }
-
-            if (file) {
-              setSvg({ data: decoder.decode(await file.getFileContents()) });
-            } else {
-              setSvg({ error: `Can't find SVG for '${props.workspaceFile.relativePath}'` });
-            }
-          });
+        // FIXME: Uncomment when KOGITO-6181 is fixed
```
Comment (tomasdavidorg) [inline packages/online-editor/src/editor/FileSwitcher.tsx:466]: Is it correct jira?[ KOGITO-6181](https://issues.redhat.com/browse/KOGITO-6181) is already resolved.

File: packages/online-editor/src/editor/FileSwitcher.tsx [hunk 4]
```diff
@@ -735,7 +734,8 @@ export function FilesMenuItems(props: {
                         </CardHeaderMain>
                       </CardHeader>
                       <CardBody style={{ padding: 0 }}>
-                        <FileSvg workspaceFile={file} />
+                        {/* FIXME: Uncomment when KOGITO-6181 is fixed */}
```
Comment (tomasdavidorg) [inline packages/online-editor/src/editor/FileSwitcher.tsx:737]: Is it correct jira?[ KOGITO-6181](https://issues.redhat.com/browse/KOGITO-6181) is already resolved.

File: packages/online-editor/src/workspace/worker/api/WorkspacesWorkerGitApi.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+ */
+
+import { WorkspaceDescriptor } from "../../model/WorkspaceDescriptor";
+import { GistOrigin, GitHubOrigin } from "../../model/WorkspaceOrigin";
+import { LocalFile } from "../../WorkspacesContext";
+import { WorkspaceWorkerFileDescriptor } from "./WorkspaceWorkerFileDescriptor";
+
+export interface WorkspacesWorkerGitApi {
+  kieSandboxWorkspacesGit_hasLocalChanges(args: {
+    workspaceId: string; //
```
Comment (tomasdavidorg) [inline packages/online-editor/src/workspace/worker/api/WorkspacesWorkerGitApi.ts:24]: Empty comment `//` + several more below in this file.

File: packages/online-editor/src/workspace/worker/api/WorkspacesWorkerStorageApi.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  ResourcesList,
+} from "@kie-tools-core/workspace/dist/api";
+import { WorkspaceWorkerFileDescriptor } from "./WorkspaceWorkerFileDescriptor";
+import { WorkspaceDescriptor } from "../../model/WorkspaceDescriptor";
+
+export interface WorkspacesWorkerStorageApi {
+  // model
+
+  kieSandboxWorkspacesStorage_deleteWorkspace(args: {
+    workspaceId: string; //
```
Comment (tomasdavidorg) [inline packages/online-editor/src/workspace/worker/api/WorkspacesWorkerStorageApi.ts:30]: Empty comment `//` + several more below in this file.

File: packages/online-editor/src/workspace/WorkspacesContext.tsx [hunk 7]
```diff
@@ -84,23 +77,27 @@ export class WorkspaceFile {
...
-  gitService: GitService;
-  svgService: WorkspaceSvgService;
-  descriptorService: WorkspaceDescriptorService;
-  fsService: WorkspaceFsService;
-
-  // create
+  // util
+
+  getUniqueFileIdentifier(args: {
+    workspaceId: string; //
```
Comment (tomasdavidorg) [inline packages/online-editor/src/workspace/WorkspacesContext.tsx:87]: Empty comment `//` + several more below in this file.

File: packages/online-editor/src/workspace/hooks/WorkspaceHooks.tsx [hunk 8]
```diff
@@ -51,15 +48,13 @@ export function useWorkspaceGitStatusPromise(workspace: ActiveWorkspace | undefi
         workspace.descriptor.origin.kind === WorkspaceKind.GIT ||
         workspace.descriptor.origin.kind === WorkspaceKind.GITHUB_GIST
       ) {
-        const head = await workspaces.gitService.resolveRef({
-          fs: await workspaces.fsService.getWorkspaceFs(workspace.descriptor.workspaceId),
-          dir: await workspaces.getAbsolutePath({ workspaceId: workspace.descriptor.workspaceId }),
+        const head = await workspaces.resolveRef({
+          workspaceId: workspace.descriptor.workspaceId, //
```
Comment (tomasdavidorg) [inline packages/online-editor/src/workspace/hooks/WorkspaceHooks.tsx:52]: Empty comment `//`.

File: packages/online-editor/src/editor/EditorPage.tsx [hunk 9]
```diff
@@ -98,77 +101,108 @@ export function EditorPage(props: Props) {
...
+          if (version + 1 < saveVersion) {
+            console.info(`Refreshing @ stale version (${version}); ignoring.`);
+            return;
+          }
+
+          console.info(`Refreshing @ current version (${saveVersion}).`);
+          refreshVersion = saveVersion;
           lastContent.current = content;
 
+          // FIXME: PMML Editor doesn't work well after this is called. Can't edit using multiple tabs.
```
Comment (tomasdavidorg) [inline packages/online-editor/src/editor/EditorPage.tsx:146]: Could we create a jira for this?

**Comment by tiagobento** [comment]
Just discovered an issue with Safari 16. Investigating.

File: packages/online-editor/src/workspace/worker/worker.ts [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+      dir: workspacesService.getAbsolutePath({ workspaceId: args.workspaceId }),
+      ref: workspace.origin.branch,
+      author: {
+        name: args.gitConfig?.name ?? GIT_USER_DEFAULT.name,
+        email: args.gitConfig?.email ?? GIT_USER_DEFAULT.email,
+      },
+      authInfo: args.authInfo,
+    });
+
+    const broadcastChannel2 = new BroadcastChannel(args.workspaceId);
```
Comment (tomasdavidorg) [inline packages/online-editor/src/workspace/worker/worker.ts:480]: The `2` suffix is not necessary as it is local variable.

**Review by tomasdavidorg — COMMENTED** [review]
Just a few comments on the code. Not yet tested.

File: packages/online-editor/src/workspace/worker/worker.ts [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+
+const GIT_USER_DEFAULT = {
+  name: "KIE Sandbox",
+  email: "",
+};
+
+async function corsProxyUrl() {
+  const envFilePath = `../../${ENV_FILE_PATH}`; // Needs to go back two dirs, since this file is at `workspaces/worker`.
+  const env = await (await fetch(envFilePath)).json();
+  return env.WEBPACK_REPLACE__corsProxyUrl ?? process.env.WEBPACK_REPLACE__corsProxyUrl ?? "";
```
Comment (caponetto) [inline packages/online-editor/src/workspace/worker/worker.ts:57]:
```suggestion
  return env.CORS_PROXY_URL ?? process.env.WEBPACK_REPLACE__corsProxyUrl ?? "";
```

**Comment by tiagobento** [comment]
I'm going to need to make more changes to it. kie-sandbox-fs doesn't work well on Safari 16. Looking for alternatives now.

Closing this one for a future PR.

**Pull request closed without merging**